forte/mrdsrg-spin-integrated/dsrg_mrpt2_grad/dsrg_mrpt2_deriv_tensors.cc

```diff
@@ -0,0 +1,150 @@
+    // X_K <K|p^+ q^+ r s|0> + <0|p^+ q^+ r s|K> X_K
+    Gamma2_tilde = BTF_->build(CoreTensor, "Gamma2_tilde", spin_cases({"aaaa"}));
+}
+
+void DSRG_MRPT2::set_density() {
+    Gamma2_ = BTF_->build(CoreTensor, "Gamma2_", spin_cases({"aaaa"}));
+
+    Gamma2_.block("aaaa")("pqrs") = rdms_->g2aa()("pqrs");
+    Gamma2_.block("aAaA")("pqrs") = rdms_->g2ab()("pqrs");
+    Gamma2_.block("AAAA")("pqrs") = rdms_->g2bb()("pqrs");
+}
+
+void DSRG_MRPT2::set_h() {
+    H = BTF_->build(CoreTensor, "One-Electron Integral", spin_cases({"gg"}));
+    H.iterate([&](const std::vector<size_t>& i, const std::vector<SpinType>& spin, double& value) {
+        if (spin[0] == AlphaSpin) {
+            value = ints_->oei_a(i[0], i[1]);
+        } else {
+            value = ints_->oei_b(i[0], i[1]);
+        }
+    });
+}
+
+void DSRG_MRPT2::set_v() {
+    if (eri_df_) {
+        // NOTICE: B block space can be further compressed.
+        B = BTF_->build(tensor_type_, "B",
+                        {"Lph", "LPH", "Lcp", "LCP", "Lav", "LAV", "Lcc", "LCC", "Lvv", "LVV"});
+        for (const std::string& block : B.block_labels()) {
+            std::vector<size_t> iaux = label_to_spacemo_[block[0]];
+            std::vector<size_t> ip = label_to_spacemo_[block[1]];
+            std::vector<size_t> ih = label_to_spacemo_[block[2]];
+            ambit::Tensor Bblock = ints_->three_integral_block(iaux, ip, ih);
+            B.block(block).copy(Bblock);
+        }
+    } else {
+        V = BTF_->build(
+            CoreTensor, "Electron Repulsion Integral",
+            spin_cases({"gphh", "pghh", "ppgh", "pphg", "gchc", "pghc", "pcgc", "pchg",
+                        "gcpc", "hgpc", "hcgc", "hcpg", "gccc", "cgcc", "ccgc", "cccg",
+                        "gcvc", "vgvc", "vcgc", "vcvg", "cgch", "gpch", "cpcg", "cpgh",
+                        "cgcp", "ghcp", "chcg", "chgp", "cgcv", "gvcv", "cvcg", "cvgv"}));
+
+        V.iterate(
+            [&](const std::vector<size_t>& i, const std::vector<SpinType>& spin, double& value) {
+                if (spin[0] == AlphaSpin) {
+                    if (spin[1] == AlphaSpin) {
+                        value = ints_->aptei_aa(i[0], i[1], i[2], i[3]);
+                    } else {
+                        value = ints_->aptei_ab(i[0], i[1], i[2], i[3]);
+                    }
+                } else if (spin[1] == BetaSpin) {
+                    value = ints_->aptei_bb(i[0], i[1], i[2], i[3]);
+                }
+            });
+    }
+
+    V_pmqm = BTF_->build(CoreTensor, "Sum_{m} V[pmqm]+ V['pMqM']", {"gg"});
+    if (eri_df_) {
+        // Summation of V["pmqm"] + V["pMqM"] over index "M"
+        V_pmqm["pq"] = 2.0 * B["gpq"] * B["gmn"] * I["mn"];
+        V_pmqm["pq"] -= B["gpm"] * B["gmq"];
+    } else {
+        // Summation of V["pmqm"] + V["pMqM"] over index "M"
+        V_pmqm["pq"] = V["pmqn"] * I["mn"];
+        V_pmqm["pq"] += V["pMqN"] * I["MN"];
+    }
+}
```

Comment:
I think these are somewhat available in the startup function. We can directly add an if statement there (i.e., if doing gradients).

Reply:
We shall discuss this later. As far as I know, the variable V in DSRG-MRPT2 is actually tilde{V}